### PR TITLE
Add `vault_alicloud_auth_backend_role` resource

### PR DESCRIPTION
**Pull request by lawliet89**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes https://github.com/terraform-providers/terraform-provider-vault/issues/664

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
* Add `vault_alicloud_auth_backend_role` resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAlicloudAuthBackendRole'

...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAlicloudAuthBackendRole -timeout 120m
?   	github.com/terraform-providers/terraform-provider-vault	[no test files]
?   	github.com/terraform-providers/terraform-provider-vault/cmd/coverage	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-vault/util	(cached) [no tests to run]
=== RUN   TestAlicloudAuthBackendRole_basic
--- PASS: TestAlicloudAuthBackendRole_basic (0.24s)
PASS
ok  	github.com/terraform-providers/terraform-provider-vault/vault	(cached)

```

### CODE REVIEW

**Comment by lawliet89** [comment]
@yongzhang  Please give this PR a try if you have the time.

**Comment by yongzhang** [comment]
@lawliet89 This is great! Thanks for supporting this!!!

File: vault/resource_alicloud_auth_backend_role.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func alicloudAuthBackendRolePath(backend, role string) string {
+	return "auth/" + strings.Trim(backend, "/") + "/role/" + strings.Trim(role, "/")
+}
+
+func alicloudAuthBackendFromPath(path string) (string, error) {
+	var parts = strings.Split(path, "/")
+	if len(parts) != 4 {
+		return "", fmt.Errorf("Expecdted 4 parts in path '%s'", path)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return "", fmt.Errorf("expected 4 parts in path '%s'", path)
```

File: vault/resource_alicloud_auth_backend_role.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	return parts[1], nil
+}
+
+func alicloudAuthRoleFromPath(path string) (string, error) {
+	var parts = strings.Split(path, "/")
+	if len(parts) != 4 {
+		return "", fmt.Errorf("Expecdted 4 parts in path '%s'", path)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return "", fmt.Errorf("expected 4 parts in path '%s'", path)
```

File: vault/resource_alicloud_auth_backend_role.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+	data := map[string]interface{}{}
+	alicloudAuthBackendRoleUpdateFields(d, data, true)
+
+	log.Printf("[DEBUG] Writing role %q to AliCloud auth backend", path)
+	d.SetId(path)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error writing AliCloud auth role %q: %s", path, err)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return fmt.Errorf("error writing AliCloud auth role %q: %s", path, err)
```

File: vault/resource_alicloud_auth_backend_role.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	data := map[string]interface{}{}
+	alicloudAuthBackendRoleUpdateFields(d, data, false)
+
+	log.Printf("[DEBUG] Updating role %q in AliCloud auth backend", path)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("Error updating AliCloud auth role %q: %s", path, err)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return fmt.Errorf("error updating AliCloud auth role %q: %s", path, err)
```

File: vault/resource_alicloud_auth_backend_role.go (hunk 1 above)
Comment (tyrannosaurus-becks) [reply vault/resource_alicloud_auth_backend_role.go]: FWIW, all errors in Go generally start with a lower-cased string.

File: vault/resource_alicloud_auth_backend_role.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func alicloudAuthBackendRoleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Reading AliCloud role %q", path)
+	resp, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("Error reading AliCloud role %q: %s", path, err)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return fmt.Errorf("error reading AliCloud role %q: %s", path, err)
```

File: vault/resource_alicloud_auth_backend_role.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func alicloudAuthBackendRoleDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Deleting AliCloud role %q", path)
+	_, err := client.Logical().Delete(path)
+	if err != nil {
+		return fmt.Errorf("Error deleting AliCloud role %q", path)
```
Comment (tyrannosaurus-becks) [inline vault/resource_alicloud_auth_backend_role.go]:
```suggestion
		return fmt.Errorf("error deleting AliCloud role %q", path)
```

**Comment by lawliet89** [comment]
Made the changes as suggested.

**Pull request merged**